File: app/src/main/java/com/sv/newsapp/activities/MapsActivity.java
package com.sv.newsapp.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import com.sv.newsapp.R;
import com.sv.newsapp.data.WeatherData;
import com.sv.newsapp.models.api.Api;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class MapsActivity extends FragmentActivity implements OnMapReadyCallback, View.OnClickListener {

    private GoogleMap mMap;

    private final Api mApi = Api.Instance.getApi();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_maps);
        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);

        findViewById(R.id.moscow_button).setOnClickListener(this);
        findViewById(R.id.spb_button).setOnClickListener(this);
        findViewById(R.id.news_button).setOnClickListener(v -> {
            Intent intent = new Intent(MapsActivity.this, MainActivity.class);
            startActivity(intent);
        });
    }

    @SuppressLint("CheckResult")
    private void receiveWeather(final String city) {
        mApi.getWeatherDataByCity(city, "d3e76d9f4759e01c064d1b34ba9ff19e",
                "metric")
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(weatherData -> showWeatherMarker(weatherData));
    }

    private void showWeatherMarker(WeatherData weatherData) {
        LatLng latLng = new LatLng(weatherData.getCoord().getLat(), weatherData.getCoord().getLon());
        mMap.addMarker(new MarkerOptions().position(latLng).title("Temp: "
                + weatherData.getMain().getTemp())).showInfoWindow();
        mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 10));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_weather, menu);

        return true;
    }

    @Override
    public void onPointerCaptureChanged(boolean hasCapture) {

    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_weather) {
            Intent intent = new Intent(this, MainActivity.class);
            startActivity(intent);
        } else if (id == R.id.action_settings) {
//
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;
    }

    @Override
    public void onClick(View v) {
        receiveWeather(((Button)v).getText().toString());
    }
}

File: app/src/main/java/com/sv/newsapp/activities/MapsActivity.kt
package com.sv.newsapp.activities

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import androidx.fragment.app.FragmentActivity
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.sv.newsapp.R
import com.sv.newsapp.data.WeatherData
import com.sv.newsapp.models.api.Api
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

class MapsActivity : FragmentActivity(), OnMapReadyCallback, View.OnClickListener {
    private var mMap: GoogleMap? = null
    private val mApi = Api.Instance.getApi()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_maps)
        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        val mapFragment = supportFragmentManager
            .findFragmentById(R.id.map) as SupportMapFragment?
        mapFragment!!.getMapAsync(this)
        findViewById<View>(R.id.moscow_button).setOnClickListener(this)
        findViewById<View>(R.id.spb_button).setOnClickListener(this)
        findViewById<View>(R.id.news_button).setOnClickListener(this)
    }

    @SuppressLint("CheckResult")
    private fun receiveWeather(city: String) {
        mApi.getWeatherDataByCity(
            city, "d3e76d9f4759e01c064d1b34ba9ff19e",
            "metric"
        )
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { weatherData: WeatherData -> showWeatherMarker(weatherData) }
    }

    private fun showWeatherMarker(weatherData: WeatherData) {
        val latLng = LatLng(weatherData.coord!!.lat.toDouble(), weatherData.coord!!.lon.toDouble())
        mMap!!.addMarker(
            MarkerOptions().position(latLng).title(
                "Temp: "
                        + weatherData.main!!.temp
            )
        ).showInfoWindow()
        mMap!!.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 10f))
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_weather, menu)
        return true
    }

    override fun onPointerCaptureChanged(hasCapture: Boolean) {}
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_weather) {
            val intent = Intent(this, MainActivity::class.java)
            startActivity(intent)
        } else if (id == R.id.action_settings) {
//
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
    }

    override fun onClick(v: View) {
        receiveWeather((v as Button).text.toString())
    }
}